ARCHI-212 Improve APIM sync process

guillaumelamirand opened the pull request:
## Issue

https://gravitee.atlassian.net/browse/ARCHI-212

<!-- Storybook placeholder -->
---

📚&nbsp;&nbsp;View the storybook of this branch [here](https://612657caa8e859003a8a6430-rcqadypxmy.chromatic.com)
<!-- Storybook placeholder end -->
<!-- UI placeholder -->
🚀 CI was able to deploy the build of this PR, so you can now try it directly [here](https://apimnightlywebui24386.z6.web.core.windows.net/archi-212/index.html)
_Notes_: The deployed app is linked to the management API of APIM master. (Same login and password as APIM master)
<!-- UI placeholder end -->

Commit:
refactor: compute latest event on management side to avoid complex searchLatest request

Commit:
fix: fix an issue while searching ids on apis with mutiple categories, labels or groups

jhaeyaert commented on gravitee-apim-gateway/gravitee-apim-gateway-services/gravitee-apim-gateway-services-sync/src/main/java/io/gravitee/gateway/services/sync/SyncManager.java:
```diff
@@ -124,45 +123,25 @@ public void startScheduler(int delay, TimeUnit unit) {
+                    } catch (Exception e) {
+                        error.set(true);
+                        lastErrorMessage = e.getMessage();
+                        logger.error("An error occurs while synchronizing organizations", e);
```
Just saw the message should be adapted also

Commit:
refactor: implement a new gateway sync process

  - make full process reactive
  - get latest events from newly events_latest collection
  - implement proper pagination mechanism to retrieve latest event
  - Retrieve in order api > plan > subscription and key before deploying anything

jhaeyaert commented on gravitee-apim-gateway/gravitee-apim-gateway-services/gravitee-apim-gateway-services-sync/src/main/java/io/gravitee/gateway/services/sync/process/fetcher/LatestEventFetcher.java:
```diff
@@ -0,0 +1,95 @@
+                        emitter.onNext(events);
+                        page.index++;
+                    } else {
+                        emitter.onComplete();
```
We could probably gain one search query by checking the result list size and complete it if it is strictly lesser than the page size. Ex:

```java
List<Event> events = eventLatestRepository.search(page.criteria, group, page.index, page.size);
      if (events != null && !events.isEmpty()) {
             emitter.onNext(events);
             page.index++;
       } 

       if(events == null || events.size() < page.size) {
              emitter.onComplete();
       }
}
```
WDYT ?

guillaumelamirand replied:
yes, in deed something I did somewhere else actually! Good catch

phiz71 commented on gravitee-apim-repository/gravitee-apim-repository-gateway-bridge-http/gravitee-apim-repository-gateway-bridge-http-client/src/main/java/io/gravitee/repository/bridge/client/management/HttpEventLatestRepository.java:
```diff
@@ -0,0 +1,77 @@
+
+    @Override
+    public Event createOrPatch(Event event) throws TechnicalException {
+        return blockingGet(post("/eventsLatest/_createOrPatch", BodyCodec.json(Event.class)).send(event)).payload();
```
Should we follow API Guidelines ?
`return blockingGet(post("/events-latest/_create-or-patch", BodyCodec.json(Event.class)).send(event)).payload();`

guillaumelamirand replied:
I didn't want to change how the bridge was built as it doesn't follow any "guidelines". Not sure it is the place to change every path, wdyt?

phiz71 replied:
I think we said that the API Guidelines have to be followed by new APIs and should follow as much as possible for existing ones.
For this bridge server API, I don't think we will one day have a project to rewrite all from scratch. So I would say, at least we could try to follow guidelines each time we can.

But maybe I'm missing something and I could understand it's not something we want to apply here.

phiz71 replied:
Ok for me. This point can be handled in a dedicated issue and PR.

phiz71 commented on gravitee-apim-gateway/gravitee-apim-gateway-services/gravitee-apim-gateway-services-sync/pom.xml:
```diff
@@ -123,6 +123,21 @@
+            <groupId>io.vertx</groupId>
+            <artifactId>vertx-junit5</artifactId>
+            <scope>test</scope>
+        </dependency>
```
duplicate

phiz71 commented on gravitee-apim-gateway/gravitee-apim-gateway-services/gravitee-apim-gateway-services-sync/src/main/java/io/gravitee/gateway/services/sync/healthcheck/SyncProcessProbe.java:
```diff
@@ -15,31 +15,32 @@
     @Override
     public String id() {
-        return "api-sync";
+        return "sync-process";
```
This has to be documented in the upgrade-guide, so that customers who rely on this probe are informed

phiz71 commented on pom.xml:
```diff
@@ -60,7 +60,7 @@
         <gravitee-expression-language.version>2.1.0</gravitee-expression-language.version>
         <gravitee-fetcher-api.version>1.4.0</gravitee-fetcher-api.version>
-        <gravitee-gateway-api.version>2.1.0</gravitee-gateway-api.version>
+        <gravitee-gateway-api.version>2.1.0-archi-212-SNAPSHOT</gravitee-gateway-api.version>
```
Do not forget to replace it, once gateway-api is released

phiz71 commented on gravitee-apim-repository/gravitee-apim-repository-test/src/test/java/io/gravitee/repository/management/ApiRepositoryTest.java:
```diff
@@ -491,13 +505,50 @@ public void searchIdsWithPageable() {
+
+        assertEquals("api-with-many-categories", apiIds.getContent().get(0));
+
+        apiIds =
+            apiRepository.searchIds(
+                List.of(new ApiCriteria.Builder().version("1").build()),
+                new PageableBuilder().pageNumber(1).pageSize(2).build(),
+                null
+            );
+
+        assertEquals(4, apiIds.getTotalElements());
+        assertEquals(2, apiIds.getPageElements());
+
+        assertEquals("api-to-update", apiIds.getContent().get(0));
+        assertEquals("grouped-api", apiIds.getContent().get(1));
+
+        apiIds =
+            apiRepository.searchIds(
+                List.of(new ApiCriteria.Builder().version("1").build()),
+                new PageableBuilder().pageNumber(0).pageSize(4).build(),
+                new SortableBuilder().field("updated_at").order(Order.DESC).build()
+            );
+
+        assertEquals("api-to-update", apiIds.getContent().get(0));
+        assertEquals("api-to-delete", apiIds.getContent().get(1));
```
Is there a link with Categories? Are these tests relevant in `shouldReturnUniqueIdsWhenSearchApisWithCategories()` method ?

guillaumelamirand replied:
The only link is that there was a bug on this and if you were searching ids using many categories, or many groups you were getting duplicated ids. One is about ids the other about API themself.

phiz71 commented on gravitee-apim-repository/gravitee-apim-repository-test/src/test/java/io/gravitee/repository/management/RepositoryTestSuite.java:
```diff
@@ -30,7 +30,7 @@
         AlertRepositoryTest.class,
         ApiHeaderTest.class,
-        ApiKeyRepositoryTest.class,
+        //        ApiKeyRepositoryTest.class,
```
Should it be uncommented?

phiz71 commented on gravitee-apim-repository/gravitee-apim-repository-test/src/test/java/io/gravitee/repository/management/SubscriptionRepositoryTest.java:
```diff
@@ -45,7 +56,7 @@ protected String getTestCasesPath() {
     public void shouldFindByPlan() throws TechnicalException {
         List<Subscription> subscriptions =
-            this.subscriptionRepository.search(new SubscriptionCriteria.Builder().plans(singleton("plan1")).build());
+            this.subscriptionRepository.search(SubscriptionCriteria.builder().plans(singleton("plan1")).build());
```
Maybe you could have put `@Singular` annotation on the `plans` field in SubscriptionCriteria class?

guillaumelamirand replied:
Will have a look.

phiz71 commented on gravitee-apim-repository/gravitee-apim-repository-test/src/test/java/io/gravitee/repository/management/SubscriptionRepositoryTest.java:
```diff
@@ -95,7 +106,7 @@ public void shouldNotFindByPlan() throws TechnicalException {
     public void shouldFindByApplication() throws TechnicalException {
         List<Subscription> subscriptions =
-            this.subscriptionRepository.search(new SubscriptionCriteria.Builder().applications(singleton("app1")).build());
+            this.subscriptionRepository.search(SubscriptionCriteria.builder().applications(singleton("app1")).build());
```
Maybe you could have put `@Singular` annotation on the `applications` field in SubscriptionCriteria class?

guillaumelamirand replied:
Will have a look.

phiz71 commented on gravitee-apim-repository/gravitee-apim-repository-jdbc/src/test/java/io/gravitee/repository/jdbc/utils/FieldUtilsTest.java:
```diff
@@ -0,0 +1,56 @@
+    public String expected;
+
+    @Test
+    public void convertToCamelCase() {
```
```suggestion
    public void convertToSnakeCase() {
```

phiz71 commented on gravitee-apim-repository/gravitee-apim-repository-gateway-bridge-http/gravitee-apim-repository-gateway-bridge-http-server/src/main/java/io/gravitee/repository/bridge/server/utils/ParamUtils.java:
```diff
@@ -0,0 +1,78 @@
+            String orderParam = ctx.request().getParam("order");
+            String fieldParam = ctx.request().getParam("field");
+            if (fieldParam != null) {
+                return new SortableBuilder().field(fieldParam).order(Order.valueOf(orderParam)).build();
```
Is there a default value if the order can't be parsed?

guillaumelamirand replied:
No default value just. an exception is thrown. Order is mandatory.

phiz71 commented on gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/rest/api/service/impl/SubscriptionServiceImpl.java:
```diff
@@ -1023,13 +1023,15 @@ public SubscriptionEntity pauseConsumer(ExecutionContext executionContext, Strin
+        streamActiveApiKeys(executionContext, subscription.getId())
+            .forEach(apiKey -> {
+                // Only paused key if the applicatio is not using shared api key
+                if (!application.hasApiKeySharedMode()) {
                     apiKey.setPaused(true);
-                    apiKeyService.update(executionContext, apiKey);
-                });
-        }
+                }
+                // Anyway update the shared key updateAt timestamp to detect changes
+                apiKeyService.update(executionContext, apiKey);
+            });
```
Maybe the method is not well named, but as far as I remember, if an application is in "shared API-key mode", it can't have some subscriptions with a shared API-key and subscriptions with "non-shared" API-key.
But happy to discuss this, as it may have changed since the last time I worked with shared keys.

guillaumelamirand replied:
we could discuss it as it was an issue on my side. Previously we weren't undeploying apikeys or subscriptions, but this was leading to apikey not undeployed.

phiz71 commented on gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/rest/api/service/impl/upgrade/EventsLatestUpgrader.java:
```diff
@@ -0,0 +1,157 @@
+            migrateEvents(Event.EventProperties.DICTIONARY_ID.getValue(), ids);
+        }
+        log.info("{} events regarding {} dictionaries have been migrated", eventsForModelCounter, modelCounter);
+    }
```
I think we should use criteria to select the rights events here.

Let's imagine that the events list for a dictionary is (from oldest to more recent one):
START_DICTIONNARY
PUBLISH_DICTIONNARY
UNPUBLISH_DICTIONNARY
PUBLISH_DICTIONNARY
STOP_DICTIONNARY

The upgrader will consider that the latest is STOP_DICTIONNARY. But the gateway should only consider PUBLISH and UNPUBLISH.

guillaumelamirand replied:
Latest event is also used on mapi side but let's discuss it with @jhaeyaert because indeed we could have an issue here if latest event is STOP, the gateway won't detect it.

phiz71 reviewed (commented):
First part of the review done (repo, rest-api, some part of the GW)
I will do the review of the gateway sync service later.

phiz71 commented on gravitee-apim-gateway/gravitee-apim-gateway-services/gravitee-apim-gateway-services-sync/src/main/java/io/gravitee/gateway/services/sync/process/handler/DefaultSyncHandler.java:
```diff
@@ -13,40 +13,40 @@
-            .put("errors", syncManager.getErrors())
-            .put("totalErrors", syncManager.getTotalErrors())
-            .put("lastErrorMessage", syncManager.getLastErrorMessage());
+            .put("initialDone", defaultSyncManager.syncDone())
+            .put("counter", defaultSyncManager.syncCounter())
+            .put("nextSyncTime", defaultSyncManager.nextSyncTime())
+            .put("lastOnError", defaultSyncManager.lastSyncOnError())
+            .put("lastErrorMessage", defaultSyncManager.lastSyncErrorMessage())
+            .put("totalOnErrors", defaultSyncManager.totalSyncOnError());
```
Same comment as for the probe name, this change has to be documented in the upgrade guide of the 4.0.0

phiz71 commented on gravitee-apim-gateway/gravitee-apim-gateway-services/gravitee-apim-gateway-services-sync/src/main/java/io/gravitee/gateway/services/sync/process/deployer/ApiKeyDeployer.java:
```diff
@@ -0,0 +1,92 @@
+                deployable
+                    .apiKeys()
+                    .stream()
+                    .sorted(comparing(ApiKey::isActive))
```
Why do we need to sort keys here ?

phiz71 commented on gravitee-apim-gateway/gravitee-apim-gateway-services/gravitee-apim-gateway-services-sync/src/main/java/io/gravitee/gateway/services/sync/process/synchronizer/organization/FlowAppender.java:
```diff
@@ -0,0 +1,67 @@
+    /**
+     * Fetching subscriptions for given deployables
+     * @param organizationDeployable the deployables to update
+     * @return the deployables updated with subscriptions
```
I think it's not the right comment

phiz71 commented on gravitee-apim-gateway/gravitee-apim-gateway-services/gravitee-apim-gateway-services-sync/src/main/java/io/gravitee/gateway/services/sync/process/synchronizer/subscription/SubscriptionSynchronizer.java:
```diff
@@ -0,0 +1,137 @@
+                        .just(subscriptions)
+                        .subscribeOn(Schedulers.from(syncFetcherExecutor))
+                        .flatMapIterable(s -> s)
+                        .filter(subscription -> planCache.isDeployed(subscription.getApi(), subscription.getPlan()))
```
Would it be possible that the plan is not yet cached?
Or maybe when the plan is cached, then its subscriptions are deployed 🤔 ?

phiz71 commented on gravitee-apim-gateway/gravitee-apim-gateway-services/gravitee-apim-gateway-services-sync/src/main/java/io/gravitee/gateway/services/sync/process/deployer/ApiDeployer.java:
```diff
@@ -0,0 +1,71 @@
+    public Completable undeploy(final ApiReactorDeployable deployable) {
+        return Completable.fromRunnable(() -> {
+            try {
+                planService.unregister(deployable);
```
when do you register the deployable?

phiz71 commented on gravitee-apim-gateway/gravitee-apim-gateway-services/gravitee-apim-gateway-services-sync/src/main/java/io/gravitee/gateway/services/sync/process/synchronizer/api/PlanAppender.java:
```diff
@@ -0,0 +1,196 @@
+        final Map<String, Api> apiById = deployables
+            .stream()
+            .map(deployable -> (Api) deployable.reactableApi())
+            .filter(api -> DefinitionVersion.V1 == api.getDefinitionVersion())
```
useless since you already did that line 56 ?

phiz71 commented on gravitee-apim-gateway/gravitee-apim-gateway-services/gravitee-apim-gateway-services-sync/src/main/java/io/gravitee/gateway/services/sync/process/synchronizer/api/PlanAppender.java:
```diff
@@ -0,0 +1,196 @@
+            .map(deployable -> {
+                ReactableApi<?> reactableApi = deployable.reactableApi();
+                if (reactableApi.getDefinitionVersion() != DefinitionVersion.V4) {
+                    filterPlanForApiV2(reactableApi);
```
this could be called for V1 API also. Is it done on purpose ?

phiz71 commented on gravitee-apim-gateway/gravitee-apim-gateway-services/gravitee-apim-gateway-services-sync/src/main/java/io/gravitee/gateway/services/sync/process/synchronizer/api/PlanAppender.java:
```diff
@@ -0,0 +1,196 @@
+        if (repoPlan.getSecurity() != null) {
+            plan.setSecurity(repoPlan.getSecurity().name());
+        } else {
+            // Handled by a migration script
```
which script ?

Commit:
refactor: change how dynamic dictionary are managed to be compatible with new process

Commit:
fix: properly manage event patch vs event latest update

Commit:
fix: reduce number of data send with hearbeat event

Commit:
fix: fix from/to criteria used to retrieve search active gateway instances

jhaeyaert commented on gravitee-apim-gateway/gravitee-apim-gateway-services/gravitee-apim-gateway-services-sync/src/main/java/io/gravitee/gateway/services/sync/process/deployer/SubscriptionDeployer.java:
```diff
@@ -0,0 +1,191 @@
+    }
+
+    @Override
+    public Completable doAtferDeployment(final SubscriptionDeployable deployable) {
```
Small typo atfer -> after

jhaeyaert commented on gravitee-apim-gateway/gravitee-apim-gateway-services/gravitee-apim-gateway-services-sync/src/main/java/io/gravitee/gateway/services/sync/healthcheck/SyncProcessProbe.java:
```diff
@@ -15,31 +15,32 @@
     @Override
     public String id() {
-        return "api-sync";
+        return "sync-process";
```
+1 for the documentation. This also has several consequences on our helm chart based on the "api-sync" probe.

guillaumelamirand replied:
Doc has been written and helm chart upgrade has been handled.

jhaeyaert commented on gravitee-apim-gateway/gravitee-apim-gateway-services/gravitee-apim-gateway-services-sync/src/main/java/io/gravitee/gateway/services/sync/process/handler/DefaultSyncHandler.java:
```diff
@@ -13,40 +13,40 @@
-            .put("errors", syncManager.getErrors())
-            .put("totalErrors", syncManager.getTotalErrors())
-            .put("lastErrorMessage", syncManager.getLastErrorMessage());
+            .put("initialDone", defaultSyncManager.syncDone())
```
Agree regarding the documentation. We should probably mention that special care must be taken for those that are relying on the "synced" attribute in their probes (I think they are using only the http status but some are also checking the body).

guillaumelamirand replied:
I think we should even say to not use this as a prob because it ignores others synchronization process. WDYT?

jhaeyaert replied:
It still makes sense to use it for the startup probe as the `DefaultSyncManager.syncDone()` moved to `true` once the initial synchronization is done (including apis, organizations, and dictionaries).

guillaumelamirand replied:
If the gateway is only using Kubernetes or the future distributed sync process, this prob wouldn't probably work.

jhaeyaert replied:
Ok, I'll resolve the conversation here but it is still something to clarify. IMHO, we still have the same need to know that everything has been synchronized when the gateway starts before accepting the traffic, whether we are getting the apis definitions from k8s CRDS or from a distributed location.

a-cordier committed:
refactor: remove dry run from apim upgraders

jhaeyaert commented on gravitee-apim-gateway/gravitee-apim-gateway-services/gravitee-apim-gateway-services-sync/src/main/java/io/gravitee/gateway/services/sync/process/synchronizer/api/SubscriptionAppender.java:
```diff
@@ -0,0 +1,104 @@
+        return deployables;
+    }
+
+    protected Map<String, List<Subscription>> loadSubscriptions(final Long frm, final Long to, final List<String> plans) {
```
frm -> from

jhaeyaert commented on gravitee-apim-gateway/gravitee-apim-gateway-services/gravitee-apim-gateway-services-sync/src/main/java/io/gravitee/gateway/services/sync/process/synchronizer/api/AbstractApiSynchronizer.java:
```diff
@@ -0,0 +1,171 @@
+                        .buffer(bulkEvents())
+                        .map(planAppender::appends)
+                        .map(deployables -> subscriptionAppender.appends(startAt, endAt, deployables))
+                        .map(deployables -> apiKeyAppender.appends(startAt, endAt, deployables))
```
I'm probably missing something regarding the `startAt` and `endAt` arguments passed to the `subscriptionAppender` and `apiKeyAppender` (certainly to optimize queries or something). Anyway, this has consequences as, once the initial sync is done, these dates are valued and applied to query the apikeys and subscription repositories but it can be too restrictive and exclude some data.

For example, let's imagine an api with an apikey plan and a valid subscription and api key. When starting the gateway and trying to reach the api, everything is working well. Stop the api, wait for a few seconds to see the api is undeployed and then start the api. The subscription and apikey have not been updated so they aren't selected by the query due to the `startAt` and `endAt` dates resulting in a 401 for the end user 💥

guillaumelamirand replied:
Fix as discussed.

jhaeyaert commented on gravitee-apim-gateway/gravitee-apim-gateway-services/gravitee-apim-gateway-services-sync/src/main/java/io/gravitee/gateway/services/sync/process/DefaultSyncManager.java:
```diff
@@ -0,0 +1,205 @@
+            .andThen(
+                Completable.fromRunnable(() -> {
+                    log.info("Sync service has been scheduled with delay [{}{}]", delay, unit.name());
+                    refreshDisposable = Flowable.interval(delay, delay, unit).flatMapCompletable(interval -> synchronize()).subscribe();
```
We should adapt this a bit to make sure that there are not several executions at the same time.` Flowable.interval` will generate a new interval for each sequence even if the  `synchronize()` isn't finished yet.

It means that, if a sync process takes more time than usual, the interval will continue to generate new intervals during that time and they will all be executed while it would be preferable to just discard them. 
For example, let's imagine a sync process every 5s, and, a synchronization taking unusual time such as 20s. Once the 20s elapsed and the synchronization completed, 4 others will be trigger immediately (but still in sequence).

guillaumelamirand replied:
Fix as discussed.

jhaeyaert commented on gravitee-apim-gateway/gravitee-apim-gateway-services/gravitee-apim-gateway-services-sync/src/main/java/io/gravitee/gateway/services/sync/process/synchronizer/api/AbstractApiSynchronizer.java:
```diff
@@ -0,0 +1,171 @@
+                ApiDeployer apiDeployer = deployerFactory.createApiDeployer();
+                return upstream
+                    .parallel(syncDeployerExecutor.getMaximumPoolSize())
+                    .runOn(Schedulers.from(syncDeployerExecutor))
```
This will switch the thread. Regarding the previous comment at the beginning of the process with the `Flowable.interval`, I think it becomes possible to have 2 executions of the sync process at the same time (and probably have consequences because of the `nextFromTime`)

guillaumelamirand replied:
The duplicate sync process has been taken care of.

Commit:
fix: improve memory consumption on bridge server/client

Commit:
fix: properly deal with backpressure on scheduled sync process

   - use Flowable.generate instead of Flowable.interval in order to properly wait for previous iteration to complete before starting a new one

Commit:
fix: properly load subscription and apikey on api incremental sync

  - previously the use of from/to for incremental update makes the process ignoring subcriptions/api keys

Commit:
quality: fix after review

Commit:
refactor: clean how ObjectMapper is used

   - disable pretty mode when writing json for database
   - try to reuse instances when possible

Commit:
Merge branch 'master' into archi-212

Commit:
chore: bump gateway-api version
